### PR TITLE
Add listeners in controllers

### PR DESCRIPTION
**Pull request by Mavhrin**

### CODE REVIEW

File: sceneformux/ux/src/main/java/com/google/ar/sceneform/ux/RotationController.java [hunk 1]
```diff
@@ -15,45 +15,62 @@
...
  * TwistGestureRecognizer}.
  */
-public class RotationController extends BaseTransformationController<TwistGesture> {
+public class RotationController extends BaseTransformationController<TwistGesture> implements MovementController {
 
-  // Rate that the node rotates in degrees per degree of twisting.
-  private float rotationRateDegrees = 2.5f;
+  public class Settings {
+    // Rate that the node rotates in degrees per degree of twisting.
+    float rotationRateDegrees = 2.5f;
```
Comment (fxsalazar) [inline sceneformux/ux/src/main/java/com/google/ar/sceneform/ux/RotationController.java]: This shouldn't still be private?

File: sceneformux/ux/src/main/java/com/google/ar/sceneform/ux/RotationController.java [hunk 2]
```diff
@@ -15,45 +15,62 @@
...
+      this.rotationRateDegrees = other.rotationRateDegrees;
+    }
   }
 
-  public void setRotationRateDegrees(float rotationRateDegrees) {
-    this.rotationRateDegrees = rotationRateDegrees;
-  }
+  public Settings settings;
+
+  @Nullable public MovementListener listener = null;
```
Comment (fxsalazar) [inline sceneformux/ux/src/main/java/com/google/ar/sceneform/ux/RotationController.java]:
```suggestion
  @Nullable 
  public MovementListener listener = null;
```

File: sceneformux/ux/src/main/java/com/google/ar/sceneform/ux/ScaleController.java [hunk 3]
```diff
@@ -15,97 +15,106 @@
...
-  private float sensitivity = DEFAULT_SENSITIVITY;
-  private float elasticity = DEFAULT_ELASTICITY;
-
-  private float currentScaleRatio;
+  private float currentScaleRatio = 0f;
 
   private static final float ELASTIC_RATIO_LIMIT = 0.8f;
   private static final float LERP_SPEED = 8.0f;
 
+  @Nullable public MovementListener listener = null;
```
Comment (fxsalazar) [inline sceneformux/ux/src/main/java/com/google/ar/sceneform/ux/ScaleController.java]:
```suggestion
  @Nullable 
  public MovementListener listener = null;
```

File: sceneformux/ux/src/main/java/com/google/ar/sceneform/ux/RotationController.java [hunk 4]
```diff
@@ -15,45 +15,62 @@
...
+  public RotationController(
+      BaseTransformableNode transformableNode, TwistGestureRecognizer gestureRecognizer) {
+    super(transformableNode, gestureRecognizer);
   }
 
   @Override
   public boolean canStartTransformation(TwistGesture gesture) {
-    return getTransformableNode().isSelected();
+    boolean selected = getTransformableNode().isSelected();
+    if (selected && null!=listener) {
```
Comment (fxsalazar) [inline sceneformux/ux/src/main/java/com/google/ar/sceneform/ux/RotationController.java]:
```suggestion
    if (selected && null != listener) {
```

File: sceneformux/ux/src/main/java/com/google/ar/sceneform/ux/RotationController.java [hunk 5]
```diff
@@ -15,45 +15,62 @@
...
-    float rotationAmount = -gesture.getDeltaRotationDegrees() * rotationRateDegrees;
+    float rotationAmount = -gesture.getDeltaRotationDegrees() * settings.rotationRateDegrees;
     Quaternion rotationDelta = new Quaternion(Vector3.up(), rotationAmount);
     Quaternion localrotation = getTransformableNode().getLocalRotation();
     localrotation = Quaternion.multiply(localrotation, rotationDelta);
-    getTransformableNode().setLocalRotation(localrotation);
+    BaseTransformableNode baseTransformableNode = getTransformableNode();
+    baseTransformableNode.setLocalRotation(localrotation);
+
+    if (null!=listener) {
```
Comment (fxsalazar) [inline sceneformux/ux/src/main/java/com/google/ar/sceneform/ux/RotationController.java]:
```suggestion
    if (null != listener) {
```

File: sceneformux/ux/src/main/java/com/google/ar/sceneform/ux/RotationController.java [hunk 6]
```diff
@@ -15,45 +15,62 @@
...
+      listener.onMovementStart(baseTransformableNode);
+    }
   }
 
   @Override
-  public void onEndTransformation(TwistGesture gesture) {}
+  public void onEndTransformation(TwistGesture gesture) {
+    if (null!=listener) {
```
Comment (fxsalazar) [inline sceneformux/ux/src/main/java/com/google/ar/sceneform/ux/RotationController.java]:
```suggestion
    if (null != listener) {
```

File: sceneformux/ux/src/main/java/com/google/ar/sceneform/ux/ScaleController.java [hunk 7]
```diff
@@ -15,97 +15,106 @@
...
       return;
     }
 
     float t = MathHelper.clamp(frameTime.getDeltaSeconds() * LERP_SPEED, 0, 1);
     currentScaleRatio = MathHelper.lerp(currentScaleRatio, getClampedScaleRatio(), t);
     float finalScaleValue = getFinalScale();
+
+    BaseTransformableNode baseTransformableNode = getTransformableNode();
+
+    if (getElasticDelta()==0f && almostEqual(baseTransformableNode.getLocalScale().x, finalScaleValue, 0.01f)) {
```
Comment (fxsalazar) [inline sceneformux/ux/src/main/java/com/google/ar/sceneform/ux/ScaleController.java]:
```suggestion
    if (getElasticDelta() == 0f && almostEqual(baseTransformableNode.getLocalScale().x, finalScaleValue, 0.01f)) {
```

File: sceneformux/ux/src/main/java/com/google/ar/sceneform/ux/ScaleController.java [hunk 8]
```diff
@@ -15,97 +15,106 @@
...
+      canUpdate = false;
+      if (null!=listener) {
```
Comment (fxsalazar) [inline sceneformux/ux/src/main/java/com/google/ar/sceneform/ux/ScaleController.java]:
```suggestion
      if (null != listener) {
```

File: sceneformux/ux/src/main/java/com/google/ar/sceneform/ux/ScaleController.java [hunk 9]
```diff
@@ -15,97 +15,106 @@
...
+    baseTransformableNode.setLocalScale(finalScale);
   }
 
   @Override
   public boolean canStartTransformation(PinchGesture gesture) {
-    return getTransformableNode().isSelected();
+    BaseTransformableNode baseTransformableNode = getTransformableNode();
+    canUpdate = baseTransformableNode.isSelected();
+    if (canUpdate) {
+      if (null!=listener) {
```
Comment (fxsalazar) [inline sceneformux/ux/src/main/java/com/google/ar/sceneform/ux/ScaleController.java]:
```suggestion
      if (null != listener) {
```

File: sceneformux/ux/src/main/java/com/google/ar/sceneform/ux/ScaleController.java [hunk 10]
```diff
@@ -115,13 +124,17 @@ public void onContinueTransformation(PinchGesture gesture) {
         || currentScaleRatio > (1.0f + ELASTIC_RATIO_LIMIT)) {
       gesture.cancel();
     }
+
+    if (null!=listener) {
```
Comment (fxsalazar) [inline sceneformux/ux/src/main/java/com/google/ar/sceneform/ux/ScaleController.java]:
```suggestion
    if (null != listener) {
```

File: sceneformux/ux/src/main/java/com/google/ar/sceneform/ux/ScaleController.java [hunk 11]
```diff
@@ -150,6 +163,10 @@ private float getElasticDelta() {
       return 0.0f;
     }
 
-    return (1.0f - (1.0f / ((Math.abs(overRatio) * elasticity) + 1.0f))) * Math.signum(overRatio);
+    return (1.0f - (1.0f / ((Math.abs(overRatio) * settings.elasticity) + 1.0f))) * Math.signum(overRatio);
+  }
+
+  private boolean almostEqual(Float value1, Float value2, Float equalityRatio) {
+    return (abs(value1) <1e-4 && abs(value2)<1e-4) || abs(value1-value2) / max(abs(value1), abs(value2)) < equalityRatio;
```
Comment (fxsalazar) [inline sceneformux/ux/src/main/java/com/google/ar/sceneform/ux/ScaleController.java]:
```suggestion
    return (abs(value1) < 1e-4 && abs(value2) < 1e-4) || abs(value1-value2) / max(abs(value1), abs(value2)) < equalityRatio;
```

File: sceneformux/ux/src/main/java/com/google/ar/sceneform/ux/TransformableNode.java [hunk 12]
```diff
@@ -26,6 +28,8 @@
 
   private final DetectedARPlanes detectedARPlanes;
 
+  @Nullable private TransformableNodeListener transformableNodeListener = null;
```
Comment (fxsalazar) [inline sceneformux/ux/src/main/java/com/google/ar/sceneform/ux/TransformableNode.java]:
```suggestion
  @Nullable
  private TransformableNodeListener transformableNodeListener = null;
```

File: sceneformux/ux/src/main/java/com/google/ar/sceneform/ux/TransformableNodeListener.java [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+
+            @Override
+            public void onMovementEnd(BaseTransformableNode baseTransformableNode) {
+                CallOnTransformableNode(baseTransformableNode, self::onScalingEnd);
+            }
+        };
+    }
+
+    public void onTranslationStart(TransformableNode transformableNode) {
+        if (null!=transformationListener) {
```
Comment (fxsalazar) [inline sceneformux/ux/src/main/java/com/google/ar/sceneform/ux/TransformableNodeListener.java]:
```suggestion
        if (null != transformationListener) {
```

File: sceneformux/ux/src/main/java/com/google/ar/sceneform/ux/TransformableNodeListener.java [hunk 14]
```diff
@@ -0,0 +1,127 @@
...
+        }
+    }
+
+    public void onScalingEnd(TransformableNode transformableNode) {
+        if (null!=transformationListener) {
+            transformationListener.onScalingEnd(transformableNode);
+        }
+    }
+
+    private void CallOnTransformableNode(BaseTransformableNode baseTransformableNode, Consumer<TransformableNode> fn) {
```
Comment (fxsalazar) [inline sceneformux/ux/src/main/java/com/google/ar/sceneform/ux/TransformableNodeListener.java]:
```suggestion
    private void callOnTransformableNode(BaseTransformableNode baseTransformableNode, Consumer<TransformableNode> fn) {
```

File: sceneformux/ux/src/main/java/com/google/ar/sceneform/ux/TransformableNodeListener.java [hunk 15]
```diff
@@ -0,0 +1,127 @@
...
+    }
+
+    public void onTranslationStart(TransformableNode transformableNode) {
+        if (null!=transformationListener) {
+            transformationListener.onTranslationStart(transformableNode);
+        }
+    }
+
+    public void onTranslationUpdate(TransformableNode transformableNode) {
+        if (null!=transformationListener) {
```
Comment (fxsalazar) [inline sceneformux/ux/src/main/java/com/google/ar/sceneform/ux/TransformableNodeListener.java]: You get the idea...

File: sceneformux/ux/src/main/java/com/google/ar/sceneform/ux/MovementListener.java [hunk 16]
```diff
@@ -0,0 +1,13 @@
+package com.google.ar.sceneform.ux;
+
+/** Interface to visual when a {@link BaseTransformableNode} is moved. */
+public interface MovementListener {
```
Comment (fxsalazar) [inline sceneformux/ux/src/main/java/com/google/ar/sceneform/ux/MovementListener.java]: I relate this `Movement` with "transalation" and maybe "rotation" but not "scaling"; It sounds more like an interaction, or action, or even transformation listener to me. I guess is the same with the movement controller.

File: sceneformux/ux/src/main/java/com/google/ar/sceneform/ux/RotationController.java (hunk 1 above)
Comment (Mavhrin) [reply sceneformux/ux/src/main/java/com/google/ar/sceneform/ux/RotationController.java]: They are public parameters that you can modify. For instance, settings of ScaleController are set to in SceneformVirtualObjectScalabilityConfiguration

File: sceneformux/ux/src/main/java/com/google/ar/sceneform/ux/RotationController.java (hunk 2 above)
Comment (Mavhrin) [reply sceneformux/ux/src/main/java/com/google/ar/sceneform/ux/RotationController.java]: done

File: sceneformux/ux/src/main/java/com/google/ar/sceneform/ux/ScaleController.java (hunk 3 above)
Comment (Mavhrin) [reply sceneformux/ux/src/main/java/com/google/ar/sceneform/ux/ScaleController.java]: done

File: sceneformux/ux/src/main/java/com/google/ar/sceneform/ux/RotationController.java (hunk 5 above)
Comment (Mavhrin) [reply sceneformux/ux/src/main/java/com/google/ar/sceneform/ux/RotationController.java]: done

File: sceneformux/ux/src/main/java/com/google/ar/sceneform/ux/TransformableNode.java (hunk 12 above)
Comment (Mavhrin) [reply sceneformux/ux/src/main/java/com/google/ar/sceneform/ux/TransformableNode.java]: done

File: sceneformux/ux/src/main/java/com/google/ar/sceneform/ux/TransformableNodeListener.java (hunk 13 above)
Comment (Mavhrin) [reply sceneformux/ux/src/main/java/com/google/ar/sceneform/ux/TransformableNodeListener.java]: done

File: sceneformux/ux/src/main/java/com/google/ar/sceneform/ux/TransformableNodeListener.java (hunk 14 above)
Comment (Mavhrin) [reply sceneformux/ux/src/main/java/com/google/ar/sceneform/ux/TransformableNodeListener.java]: done

File: sceneformux/ux/src/main/java/com/google/ar/sceneform/ux/TransformableNodeListener.java (hunk 15 above)
Comment (Mavhrin) [reply sceneformux/ux/src/main/java/com/google/ar/sceneform/ux/TransformableNodeListener.java]: done

**Pull request merged**